pkg/restore/batcher.go

```diff
@@ -0,0 +1,252 @@
+	sender BatchSender,
+	manager ContextManager,
+	errCh chan<- error,
+) (*Batcher, <-chan CreatedTable) {
+	output := make(chan CreatedTable, defaultBatcherOutputChannelSize)
+	b := &Batcher{
+		rewriteRules:       EmptyRewriteRule(),
+		sendErr:            errCh,
+		outCh:              output,
+		sender:             sender,
+		manager:            manager,
+		cachedTablesMu:     new(sync.Mutex),
+		batchSizeThreshold: 1,
+	}
+	return b, output
+}
+
+// EnableAutoCommit enables the batcher commit batch periodicity even batcher size isn't big enough.
+// we make this function for disable AutoCommit in some case.
+func (b *Batcher) EnableAutoCommit(ctx context.Context, delay time.Duration) {
+	if b.joiner != nil {
+		log.Warn("enable auto commit on a batcher that is enabled auto commit, nothing will happen")
+		log.Info("if desire, please disable auto commit firstly")
+	}
+	joiner := make(chan struct{})
+	go b.workLoop(ctx, joiner, delay)
+	b.joiner = joiner
+}
+
+// DisableAutoCommit blocks the current goroutine until the worker can gracefully stop,
+// and then disable auto commit.
+func (b *Batcher) DisableAutoCommit(ctx context.Context) {
+	b.joinWorker()
+	b.joiner = nil
+}
+
+// joinWorker blocks the current goroutine until the worker can gracefully stop.
+// return immediately when auto commit disabled.
+func (b *Batcher) joinWorker() {
+	if b.joiner != nil {
+		log.Debug("gracefully stoping worker goroutine")
+		b.joiner <- struct{}{}
+		log.Debug("gracefully stopped worker goroutine")
+	}
+}
+
+func (b *Batcher) workLoop(ctx context.Context, joiner <-chan struct{}, delay time.Duration) {
+	tick := time.NewTicker(delay)
+	defer tick.Stop()
+	for {
+		select {
+		case <-joiner:
+			log.Debug("graceful stop signal received")
+			return
+		case <-ctx.Done():
+			b.sendErr <- ctx.Err()
+			return
+		case <-tick.C:
+			if b.Len() > 0 {
+				log.Info("sending batch because time limit exceed", zap.Int("size", b.Len()))
+				b.asyncSend(ctx)
+			}
+		}
+	}
+}
+
+func (b *Batcher) asyncSend(ctx context.Context) {
+	tbls, err := b.Send(ctx)
+	if err != nil {
+		b.sendErr <- err
+		return
+	}
+	for _, t := range tbls {
+		b.outCh <- t
+	}
+}
+
+type drainResult struct {
+	// TablesToSend are tables that would be send at this batch.
+	TablesToSend []CreatedTable
+	// BlankTablesAfterSend are tables that will be full-restored after this batch send.
+	BlankTablesAfterSend []CreatedTable
+	RewriteRules         *RewriteRules
+	Ranges               []rtree.Range
+}
+
+func newDrainResult() drainResult {
+	return drainResult{
+		TablesToSend:         make([]CreatedTable, 0),
+		BlankTablesAfterSend: make([]CreatedTable, 0),
+		RewriteRules:         EmptyRewriteRule(),
+		Ranges:               make([]rtree.Range, 0),
+	}
+}
+
+func (b *Batcher) drainRanges() drainResult {
+	result := newDrainResult()
+
+	b.cachedTablesMu.Lock()
+	defer b.cachedTablesMu.Unlock()
+
+	for offset, thisTable := range b.cachedTables {
+		thisTableLen := len(thisTable.Range)
```

Comment:
It seems we can skip empty thisTable range here, or it will log empty table repeatedly in [L198](https://github.com/pingcap/br/pull/266/files#diff-73cab7148a1a613ea6e6fe219e2b6b89R198) and do SplitPreWork repeatedly in [L205](https://github.com/pingcap/br/pull/266/files#diff-73cab7148a1a613ea6e6fe219e2b6b89R205)